src/main.vue

```diff
@@ -0,0 +1,418 @@
+import '@/components/utils/InputPrompt';
+import '@/components/utils/InputConfirm';
+
+let logLevelMatch = window.location.href.match(/kiwi-loglevel=(\d)/);
+if (logLevelMatch && logLevelMatch[1]) {
+    Logger.setLevel(parseInt(logLevelMatch[1], 10));
+}
+
+let log = Logger.namespace('main');
+
+// Add the global API as soon as possible so that things can start listening to it
+let api = window.kiwi = GlobalApi.singleton();
+
+// Third party imports now have access to the state and api
+/* eslint-disable import/first */
+import '@/thirdparty/';
+
+function getQueryVariable(variable) {
+    let query = window.location.search.substring(1);
+    let vars = query.split('&');
+    for (let i = 0; i < vars.length; i++) {
+        let pair = vars[i].split('=');
+        if (pair[0] === variable) {
+            return pair[1];
+        }
+    }
+
+    return false;
+}
+
+// Add a handy this.listen() fn to Vue instances. Saves on the need to add an event listener
+// and then manually remove them all the time.
+Vue.mixin({
+    beforeDestroy: function beforeDestroy() {
+        (this.listeningEvents || []).forEach(fn => fn());
+    },
+    methods: {
+        listen: function listen(source, event, fn) {
+            this.listeningEvents = this.listeningEvents || [];
+            this.listeningEvents.push(() => {
+                (source.$off || source.off).call(source, event, fn);
+            });
+            (source.$on || source.on).call(source, event, fn);
+        },
+        listenOnce: function listenOnce(source, event, fn) {
+            this.listeningEvents = this.listeningEvents || [];
+            this.listeningEvents.push(() => {
+                (source.$off || source.off).call(source, event, fn);
+            });
+            (source.$once || source.once).call(source, event, fn);
+        },
+    },
+});
+
+// Make the state available to all components by default
+Vue.mixin({
+    computed: {
+        $state() {
+            return state;
+        },
+    },
+});
+
+// Allow adding existing raw elements to component templates
+// Eg: <div v-rawElement="domElement"></div>
+// Eg: <div v-rawElement="{el: domElement, data:{foo:'bar'}}"></div>
+Vue.directive('rawElement', {
+    bind(el, binding) {
+        if (binding.value.nodeName) {
+            el.appendChild(binding.value);
+        } else if (binding.value.el) {
+            let rawEl = binding.value.el;
+            el.appendChild(rawEl);
+
+            // Add any data attributes to the raw element
+            if (binding.value.data) {
+                Object.keys(binding.value.data).forEach((key) => {
+                    rawEl.dataset[key] = binding.value.data[key];
+                });
+            }
+
+            // Add any properties to the raw element
+            if (binding.value.props) {
+                Object.keys(binding.value.props).forEach((key) => {
+                    rawEl[key] = binding.value.props[key];
+                });
+            }
+        }
+    },
+});
```

Comment:
Do these vue mixins and directives get added to the parent application? This could cause conflicts

Reply:
I think they do, but there's not really a way to prevent that. Perhaps a warning in the docs would suffice?